tests/test-class-wc-payment-gateway-wcpay-subscriptions.php

```diff
@@ -171,6 +171,93 @@ public function test_update_failing_payment_method_does_not_copy_method_if_token
+		$subscription->update_meta_data( '_old_payment_method_title', $old_payment_method_title );
+		$subscription->set_payment_method( $new_payment_method );
+		$subscription->set_payment_method_title( $new_payment_method_title );
+
+		$post_payment_token_parameter  = 'wc-' . WC_Payment_Gateway_WCPay::GATEWAY_ID . '-payment-token';
+		$post_payment_method_parameter = 'wcpay-payment-method';
+
+		$_POST[ $post_payment_token_parameter ] = $token2->get_id();
+
+		$old_payment_method_title_modified = (string) apply_filters( 'woocommerce_subscription_note_old_payment_method_title', $old_payment_method_title, $old_payment_method, $subscription );
+		$new_payment_method_title_modified = (string) apply_filters( 'woocommerce_subscription_note_new_payment_method_title', $new_payment_method_title, $new_payment_method, $subscription );
+		$this->assertContains( $last4_1, $old_payment_method_title_modified );
+		$this->assertContains( $last4_2, $new_payment_method_title_modified );
+
+		// case 2: same payment method, different last 4, new payment is *not* saved. expect api call called once and both old and new title to be modified.
+		$payment_method_id                       = 'test-payment-method-id';
+		$_POST[ $post_payment_token_parameter ]  = '';
+		$_POST[ $post_payment_method_parameter ] = $payment_method_id;
+		$this->mock_api_client
+			->expects( $this->once() )
+			->method( 'get_payment_method' )
+			->with( $payment_method_id )
+			->willReturn(
+				[
+					'card' => [
+						'last4' => $last4_3,
+					],
+				]
+			);
+		$old_payment_method_title_modified = (string) apply_filters( 'woocommerce_subscription_note_old_payment_method_title', $old_payment_method_title, $old_payment_method, $subscription );
+		$new_payment_method_title_modified = (string) apply_filters( 'woocommerce_subscription_note_new_payment_method_title', $new_payment_method_title, $new_payment_method, $subscription );
+		$this->assertContains( $last4_1, $old_payment_method_title_modified );
+		$this->assertContains( $last4_3, $new_payment_method_title_modified );
+
+		// case 3: non wcpay. expect both old and new title not modified.
+		$old_payment_method = 'non-wc-pay';
+		$new_payment_method = 'non-wc-pay';
+		$subscription->update_meta_data( '_old_payment_method', $old_payment_method );
+		$subscription->set_payment_method( $new_payment_method );
+		$old_payment_method_title_modified = (string) apply_filters( 'woocommerce_subscription_note_old_payment_method_title', $old_payment_method_title, $old_payment_method, $subscription );
+		$new_payment_method_title_modified = (string) apply_filters( 'woocommerce_subscription_note_new_payment_method_title', $new_payment_method_title, $new_payment_method, $subscription );
+		$this->assertEquals( $old_payment_method_title, $old_payment_method_title_modified );
+		$this->assertEquals( $new_payment_method_title, $new_payment_method_title_modified );
+
+		// case 4: different payment method. expect both old and new title not modified.
```

Comment:
Mentioned above, I'm not 100% sure about this cross GW case.

Reply:
What is GW?

I put answers above about the behaviour of old and new hook depends on new and old payment, respectively.
I assume that that business logic.

Reply:
sorry, GW = gateway.

Reply:
Point taken. cross GW case has been removed.